libmamba/src/core/transaction.cpp

```diff
@@ -825,208 +828,231 @@ namespace mamba
+                true,
+                [](bool acc, const auto& r) { return acc && r; }
+            );
+            return all_downloaded;
         }
-        // make sure that all targets have finished extracting
-        while (!is_sig_interrupted())
+
+        bool check_all_valid(const FetcherList& fetchers, const ExtractRequestList& extract_requests)
         {
-            bool all_finished = true;
-            for (const auto& t : targets)
+            bool all_valid = true;
+            for (auto [fit, eit] = std::tuple{ fetchers.begin(), extract_requests.begin() };
+                 eit != extract_requests.end();
+                 ++fit, ++eit)
             {
-                if (!t->finished())
+                PackageExtractRequest::Result res = eit->get_result();
+                if (!res.valid || !res.extracted)
                 {
-                    all_finished = false;
-                    break;
+                    fit->clear_cache();
+                    all_valid = false;
+                    // TODO: check if we can remove this
+                    throw std::runtime_error(
+                        std::string("Found incorrect download: ") + fit->name() + ". Aborting"
+                    );
                 }
             }
-            if (all_finished)
-            {
-                break;
-            }
-            std::this_thread::sleep_for(std::chrono::milliseconds(100));
+            return all_valid;
+        }
+    }
+
+    bool MTransaction::fetch_extract_packages()
+    {
+        // TODO: move this to the PackageDownloadMonitor
+        auto& pbar_manager = Console::instance().init_progress_bar_manager(ProgressBarMode::aggregated
+        );
+
+        auto& channel_context = m_pool.channel_context();
+        auto& ctx = channel_context.context();
+        PackageFetcherSemaphore::set_max(ctx.threads_params.extract_threads);
+
+        FetcherList fetchers = build_fetchers(m_pool, m_solution, m_multi_cache);
+
+        auto download_end = std::partition(
+            fetchers.begin(),
+            fetchers.end(),
+            [](const auto& f) { return f.needs_download(); }
+        );
+        auto extract_end = std::partition(
+            download_end,
+            fetchers.end(),
+            [](const auto& f) { return f.needs_extract(); }
+        );
+
+        auto download_size = static_cast<std::size_t>(std::distance(fetchers.begin(), download_end));
+        auto extract_size = static_cast<std::size_t>(std::distance(fetchers.begin(), extract_end));
+
+        // At this point:
+        // - [fetchers.begin(), download_end) contains packages that need to be downloaded,
+        // validated and extracted
+        // - [download_end, extract_end) contains packages that need to be extracted only
+        // - [extract_end, fecthers.end()) contains packages already installed and extracted
+
+        ExtractRequestList extract_requests = build_extract_requests(ctx, fetchers, extract_size);
```

Comment:
I think this can be `const` now?

Reply:
Not really, we call non const method of the extract request later.